minima/log.nim

```diff
@@ -0,0 +1,125 @@
+# Minima
+#
+# Copyright (c) 2020 Dean Eigenmann
+# Licensed under MIT license ([LICENSE](LICENSE) or http://opensource.org/licenses/MIT)
+
+import sequtils, stew/endians2, nimcrypto, random
+
+type 
+    ## Log represents the the Write-Ahead Log used for persisting the KV Database.
+    Log* = ref object of RootObj
+        file: File
+
+    StandardLog* = ref object of Log
+    EncryptedLog* = ref object of Log
+        key: array[aes256.sizeKey, byte]
+
+method close*(log: Log) {.base.} =
+    ## Closes the log file.
+    log.file.close()
+
+method log*(log: Log, key: seq[byte], value: seq[byte]) {.base.} =
+    ## Writes a key, value pair to the Write-Ahead Log.
+    discard
+
+method next*(log: Log): (seq[byte], seq[byte]) {.base.} =
+    ## Next gets the next key, value pair from the log.
+    discard
+
+iterator pairs*(log: Log): (seq[byte], seq[byte]) =
+    ## Pairs is an iterator that iterates through all the pairs in the Write-Ahead Log.
+    while log.file.getFilePos() <= log.file.getFileSize() - 1:
+        var (k, v) = log.next()
+        yield (k, v)
+
+proc readInt(file: File): int =
+    var arr: array[4, byte]
+    discard file.readBytes(arr, 0, 4)
+
+    return int(uint32.fromBytes(arr))
+
+proc pack(key, value: seq[byte]): seq[byte] =
+    concat(
+        @(uint32(len(key)).toBytes),
+        @(uint32(len(value)).toBytes),
+        key,
+        value
+    )
+
+proc init*(T: type StandardLog, file: File): T =
+    ## Returns a new Log.
+    result = T(file: file)
+
+method log*(log: StandardLog, key: seq[byte], value: seq[byte]) =
+    let write = pack(key, value)
+    # @TODO CATCH EXCEPTIONS
```

Comment:
One can use the `raises` pragma to enforce at compile time such exception-safety and specify which specific exceptions can be thrown, so that one can throw only those that are part of an API intended: https://nim-lang.org/docs/manual.html#effect-system-exception-tracking